src/transform/src/typecheck.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Check that the visible type of each query has not been changed
+
+use std::{cell::RefCell, collections::BTreeMap};
+
+use mz_compute_client::types::dataflows::BuildDesc;
+use mz_expr::{Id, OptimizedMirRelationExpr};
+use mz_repr::ColumnType;
+use tracing::{error, info, warn};
+
+use crate::TransformError;
+
+/// Check that the visible type of each query has not been changed
+#[derive(Debug)]
+pub struct Typecheck {
+    /// The known types of the queries so far
+    pub ctx: RefCell<BTreeMap<Id, Vec<ColumnType>>>,
+}
+
+impl Default for Typecheck {
+    fn default() -> Self {
+        Self {
+            ctx: RefCell::new(BTreeMap::new()),
+        }
+    }
+}
+
+impl crate::Transform for Typecheck {
+    #[tracing::instrument(
+        target = "optimizer"
+        level = "trace",
+        skip_all,
+        fields(path.segment = "typecheck")
+    )]
+    fn transform(
+        &self,
+        relation: &mut mz_expr::MirRelationExpr,
+        _args: crate::TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        let ctx = self.ctx.borrow();
+
+        if let Err(err) = relation.typecheck(&ctx) {
+            warn!("TYPE ERROR: {}\nIN UNKNOWN QUERY:\n{:#?}", err, relation);
+        }
+
+        Ok(())
+    }
+
+    #[tracing::instrument(
+        target = "optimizer"
+        level = "trace",
+        skip_all,
+        fields(path.segment = "typecheck")
+    )]
+    fn transform_query(
+        &self,
+        build_desc: &mut BuildDesc<OptimizedMirRelationExpr>,
+        _args: crate::TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        let BuildDesc { id, plan } = build_desc;
+        let mut ctx = self.ctx.borrow_mut();
+
+        let expected = ctx.get(&Id::Global(*id));
+
+        if expected.is_none() && !id.is_transient() {
+            info!(
+                "TYPECHECKER FOUND NEW NON-TRANSIENT TOP LEVEL QUERY {}\n{:#?}",
+                id, plan
+            );
```

Comment:
This should probably be an error?

Reply:
This will always break on the first run, then: you can't have an entry for a global until you've met it. (This may change when we pull global information from the `QueryContext`.

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: internal transform error: FOUND NON-TRANDSIENT TOP LEVEL QUERY BOUND TO s415:
ArrangeBy keys=[[#2]]
  Get s298


Stack backtrace:
   0: mz_environmentd::serve::{closure#0}::{closure#0}
   1: <tokio::runtime::park::CachedParkThread>::block_on::<mz_environmentd::serve::{closure#0}>
   2: <tokio::runtime::scheduler::multi_thread::MultiThread>::block_on::<mz_environmentd::serve::{closure#0}>
   3: <tokio::runtime::runtime::Runtime>::block_on::<mz_environmentd::serve::{closure#0}>
   4: std::sys_common::backtrace::__rust_begin_short_backtrace::<<mz_sqllogictest::runner::RunnerInner>::start::{closure#0}::{closure#3}, ()>
   5: <<std::thread::Builder>::spawn_unchecked_<<mz_sqllogictest::runner::RunnerInner>::start::{closure#0}::{closure#3}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
   6: std::sys::unix::thread::Thread::new::thread_start
   7: <unknown>
   8: __clone', src/sqllogictest/src/bin/sqllogictest.rs:168:51
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: sqllogictest::main::{closure#0}
   4: <tokio::runtime::park::CachedParkThread>::block_on::<sqllogictest::main::{closure#0}>
   5: <tokio::runtime::scheduler::multi_thread::MultiThread>::block_on::<sqllogictest::main::{closure#0}>
   6: <tokio::runtime::runtime::Runtime>::block_on::<sqllogictest::main::{closure#0}>
   7: sqllogictest::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
ci/test/slt-fast.sh: line 162:     8 Aborted                 sqllogictest -v "${tests[@]}" "$@"
```

I made it so this is an error only after the initial run, though it pollutes the API a little bit: the _first_ instance of `Typecheck` should allow fresh global IDs, but later ones should. Mostly people are running `logical_optimizer()`, but some clients only use `logical_cleanup_pass()`. I added a parameter, but things are starting to look gross. 😒